File: room-service/src/main/java/dev/arkaan/schoolapp/roomservice/DBConfiguration.java
package dev.arkaan.schoolapp.roomservice;

import com.zaxxer.hikari.HikariDataSource;
import org.jdbi.v3.core.Jdbi;
import org.springframework.boot.context.properties.ConfigurationProperties;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

import javax.sql.DataSource;

@Configuration
public class DBConfiguration {

    @Bean
    @ConfigurationProperties("spring.datasource")
    public DataSource dataSource() {
        return new HikariDataSource();
    }

    @Bean
    public Jdbi jdbi(DataSource dataSource) {
        return Jdbi.create(dataSource);
    }
}

File: subject-service/src/main/kotlin/dev/arkaan/schoolapp/subjectservice/db/DBConfiguration.kt
package dev.arkaan.schoolapp.subjectservice.db

import com.zaxxer.hikari.HikariDataSource
import io.micronaut.context.annotation.ConfigurationProperties
import io.micronaut.context.annotation.Factory
import jakarta.inject.Named
import jakarta.inject.Singleton
import org.jetbrains.exposed.sql.Database
import javax.sql.DataSource

@ConfigurationProperties("datasource.subject")
data class DBConfiguration(
    val jdbcUrl: String,
    val username: String,
    val password: String,
    val dialect: String,
    val driverClassName: String
)

@Factory
class MyDataSource {

    @Singleton
    @Named("subject")
    fun datasource(config: DBConfiguration): DataSource {
        println(config)

        return HikariDataSource().apply {
            username = config.username
            password = config.password
            jdbcUrl = config.jdbcUrl
            driverClassName = config.driverClassName
            maximumPoolSize = 4
        }
    }

    @Singleton
    @Named("subject")
    fun exposed(@Named("subject") dataSource: DataSource): Database {
        val db by lazy {
            Database.connect(dataSource)
        }
        return db
    }
}